idx/cassandra/cassandra.go

```diff
@@ -0,0 +1,243 @@
+package cassandra
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/metrictank/idx/memory"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+const keyspace_schema = `CREATE KEYSPACE IF NOT EXISTS %s WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}  AND durable_writes = true`
+const table_schema = `CREATE TABLE IF NOT EXISTS %s.metric_def_idx (
+    id text PRIMARY KEY,
+    def blob,
+) WITH compaction = {'class': 'SizeTieredCompactionStrategy'}
+    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}`
+
+var (
+	idxCasOk             met.Count
+	idxCasFail           met.Count
+	idxCasAddDuration    met.Timer
+	idxCasDeleteDuration met.Timer
+
+	Enabled        bool
+	keyspace       string
+	hosts          string
+	consistency    string
+	timeout        time.Duration
+	numConns       int
+	writeQueueSize int
+	protoVer       int
+)
+
+func ConfigSetup() {
+	casIdx := flag.NewFlagSet("cassandra-idx", flag.ExitOnError)
+	casIdx.BoolVar(&Enabled, "enabled", false, "")
+	casIdx.StringVar(&keyspace, "keyspace", "metric", "Cassandra keyspace to store metricDefinitions in.")
+	casIdx.StringVar(&hosts, "hosts", "localhost:9042", "comma separated list of cassandra addresses in host:port form")
+	casIdx.StringVar(&consistency, "consistency", "one", "write consistency (any|one|two|three|quorum|all|local_quorum|each_quorum|local_one")
+	casIdx.DurationVar(&timeout, "timeout", time.Second, "cassandra request timeout")
+	casIdx.IntVar(&numConns, "num-conns", 10, "number of concurrent connections to cassandra")
+	casIdx.IntVar(&writeQueueSize, "write-queue-size", 100000, "Max number of metricDefs allowed to be unwritten to cassandra")
+	casIdx.IntVar(&protoVer, "protocol-version", 4, "cql protocol version to use")
+	globalconf.Register("cassandra-idx", casIdx)
+}
+
+type writeReq struct {
+	def      *schema.MetricDefinition
```

Comment:
it doesnt need to be a pointer, but a pointer here performs better and uses about 3% less memory per metricDef and 1 less allocation.

```
#PTR
BenchmarkIndexing-4	   30000	     54006 ns/op	    4447 B/op	      72 allocs/op
#NOT PTR
BenchmarkIndexing-4	   30000	     54876 ns/op	    4592 B/op	      73 allocs/op
```
given that these should not be in the writeQueue long, they shouldnt impact GC very much.

We dont need to avoid using pointers everywhere, just in Slices, as slices are passed around by reference anyway, so there is no memory saving by passing a slice of pointers over a slice of values.